src/runtime_src/core/pcie/driver/linux/xocl/subdev/firewall.c

```diff
@@ -151,48 +151,46 @@ struct firewall_ip {
 
 struct firewall {
 	struct firewall_ip	af[MAX_LEVEL];
-	u32			max_level;
-
-	u32			curr_status;
-	int			curr_level;
-
-	u32			err_detected_status;
-	u32			err_detected_level;
-	u64			err_detected_time;
-	u64			err_detected_araddr;
-	u64			err_detected_awaddr;
-	u32			err_detected_aruser;
-	u32			err_detected_awuser;
-
-	bool			inject_firewall;
-	u64			cache_expire_secs;
-	struct xcl_firewall	cache;
-	ktime_t			cache_expires;
-	char		err_detected_level_name[50];
-	char		level_name[MAX_LEVEL][50];
+	struct xcl_firewall	status;
+	ktime_t				last_update;
+	ktime_t				expire_secs;
+	char				level_name[MAX_LEVEL][50];
+
+	bool				inject_firewall;
+	u64					err_detected_araddr;
+	u64					err_detected_awaddr;
+	u32					err_detected_aruser;
+	u32					err_detected_awuser;
 };
 
 static int clear_firewall(struct platform_device *pdev);
 static u32 check_firewall(struct platform_device *pdev, int *level);
 
-static void set_fw_data(struct firewall *fw, struct xcl_firewall *fw_status)
-{
-	memcpy(&fw->cache, fw_status, sizeof(struct xcl_firewall));
-	fw->cache_expires = ktime_add(ktime_get_boottime(),
-		ktime_set(fw->cache_expire_secs, 0));
-}
-
-static void fw_read_from_peer(struct platform_device *pdev)
+/* 
+ * Request the firewall status from the mgmt driver via mailbox.
+ * Populates the device firewall status struct with the response
+ * from the mgmt driver
+ */
+static void request_firewall_status(struct platform_device *pdev)
 {
 	struct firewall *fw = platform_get_drvdata(pdev);
 	struct xcl_mailbox_subdev_peer subdev_peer = {0};
-	struct xcl_firewall fw_status = {0};
 	size_t resp_len = sizeof(struct xcl_firewall);
 	size_t data_len = sizeof(struct xcl_mailbox_subdev_peer);
 	struct xcl_mailbox_req *mb_req = NULL;
 	size_t reqlen = sizeof(struct xcl_mailbox_req) + data_len;
 	xdev_handle_t xdev = xocl_get_xdev(pdev);
 
+	/* 
+	 * If the last status request has not expired
+	 * use the previous status update
+	 */
+	const ktime_t now = ktime_get_boottime();
+	if (ktime_compare(now, fw->last_update) < 0)
+		return;
+
+	fw->last_update = ktime_add(now, fw->expire_secs);
+
```

Comment:
We should only call this function when receiving firewall trip notification. And we need request firewall data from the remote peer anyway. We do not need to check the expire time.

Reply:
The get_prop function is written to perform a separate mailbox request for each invocation. This causes a lot of dmesg logging regarding the firewall. By adding the expiration timer it prevents a series of mailbox requests for firewall information.